src/EditorFeatures/Core/Intents/IIntentProcessor.cs

```diff
@@ -0,0 +1,85 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Features.Intents;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.Text;
+
+namespace Microsoft.CodeAnalysis.Editor.Intents
+{
+    internal interface IIntentProcessor
+    {
+        /// <summary>
+        /// For an input intent, computes the edits required to apply that intent and returns them.
+        /// </summary>
+        /// <param name="intentRequestContext">the intents with the context in which the intent was found.</param>
+        /// <returns>the edits that should be applied to the current snapshot.</returns>
+        Task<IntentResult?> ComputeEditsAsync(IntentRequestContext intentRequestContext, CancellationToken cancellationToken);
+    }
+
+    /// <summary>
+    /// Defines the data needed to compute the code action edits from an intent.
+    /// </summary>
+    internal struct IntentRequestContext
+    {
+        /// <summary>
+        /// The intent name.  <see cref="WellKnownIntents"/> contains all intents roslyn knows how to handle.
+        /// </summary>
+        public string IntentName { get; }
+
+        /// <summary>
+        /// JSON formatted data specific to the intent that must be deserialized into the appropriate object.
+        /// </summary>
+        public string? IntentData { get; }
+
+        /// <summary>
+        /// The text and selection in the document before changes were made triggering an intent.
+        /// <remarks>
+        /// For example, if the text typed was 'public Class(' which led to a generate ctor intent,
+        /// we expect a snapshot before any of the ctor was typed.
+        /// </remarks>
+        /// </summary>
+        public SnapshotSpan PriorSnapshotSpan { get; }
```

Comment:
@intellicode

4.  For `PriorSnapshotSpan` and `CurrentSnapshotSpan`, are you able to provide these as actual snapshot spans?  We discussed returning the edits required to map back to the prior, but it seems like either of us could rewind (unless you are not able to get the actual snapshot).  We're open to almost any combination (give us prior + forward edits, give us current + backward edits, give us snapshots for both), I just chose this initially as having both allows us to re-use caches.
5. Regardless of 4) above, are you able to provide us the sequence of forward text edits?  We may want to use them to be smarter on generating the text edits eventually (e.g. person types private, so we generate a private ctor).

Reply:
@dibarbet  lets sync on this with some of our code at hand, than I can show what we have and you can tell how we transform that to the types you prefer

I think it will be something like the current + how to come to the initial state (backward edit)